$GLOBALS['TYPO3_DB'] in OaiPmh.php

erik-konrad opened the pull request:
refactores $GLOBALS['TYPO3_DB'] to Doctrine/Dbal plain SQL calls. The use of the QueryBuilder is not possible because the QueryBuilder doesn't support GROUP_CONCAT (by the way, GROUP_CONCAT is not very compatible with e.g. PostgreSQL).

Commit:
[9LTS] OaiPmh.php refactoring

Commit:
Merge branch 'master' of https://github.com/erik-konrad/kitodo-presentation

Commit:
removes whitespace before comma

sebastian-meyer committed:
Merge branch 'master' into master

sebastian-meyer committed:
Fix some code styling

sebastian-meyer commented on Classes/Plugin/OaiPmh.php:
```diff
@@ -449,35 +456,43 @@ protected function resume()
+            'FROM `tx_dlf_documents` ' .
+            'INNER JOIN `tx_dlf_relations` ON `tx_dlf_relations`.`uid_local` = `tx_dlf_documents`.`uid` ' .
+            'INNER JOIN `tx_dlf_collections` ON `tx_dlf_collections`.`uid` = `tx_dlf_relations`.`uid_foreign` ' .
+            'WHERE (((`tx_dlf_documents`.`deleted` = 0) AND (`tx_dlf_collections`.`deleted` = 0)) AND ((`tx_dlf_documents`.`hidden` = 0) AND (`tx_dlf_collections`.`hidden` = 0)) AND (`tx_dlf_documents`.`starttime` <= 1586953440) AND ((`tx_dlf_documents`.`endtime` = 0) OR (`tx_dlf_documents`.`endtime` > 1586953440))) ' .
```
Values for `tx_dlf_documents.starttime` and `tx_dlf_documents.endtime` should be dynamically depending on `$GLOBALS['EXEC_TIME']`.

erik-konrad replied:
That was copied from debug(). I fix it.

sebastian-meyer replied:
We should use the helper function `Helper::whereExpression()` here.

sebastian-meyer commented on Classes/Plugin/OaiPmh.php:
```diff
@@ -931,24 +946,32 @@ protected function fetchDocumentUIDs()
+            'FROM `tx_dlf_documents` ' .
+            'INNER JOIN `tx_dlf_relations` ON `tx_dlf_relations`.`uid_local` = `tx_dlf_documents`.`uid` ' .
+            'INNER JOIN `tx_dlf_collections` ON `tx_dlf_collections`.`uid` = `tx_dlf_relations`.`uid_foreign` ' .
+            'WHERE (((`tx_dlf_documents`.`deleted` = 0) AND (`tx_dlf_collections`.`deleted` = 0)) AND ((`tx_dlf_documents`.`hidden` = 0) AND (`tx_dlf_collections`.`hidden` = 0)) AND (`tx_dlf_documents`.`starttime` <= 1586953440) AND ((`tx_dlf_documents`.`endtime` = 0) OR (`tx_dlf_documents`.`endtime` > 1586953440))) ' .
```
Values for `tx_dlf_documents.starttime` and `tx_dlf_documents.endtime` should be dynamically depending on `$GLOBALS['EXEC_TIME']`.

sebastian-meyer commented on Classes/Plugin/OaiPmh.php:
```diff
@@ -1018,18 +1041,21 @@ protected function generateOutputForDocumentList(DocumentList $documentListSet)
-                'tx_dlf_tokens',
-                [
+        if ($documentListSet->count() !== 0) {
+            $token = uniqid('', false);
```
Is this just for clarity or why do you explicitly set the default parameters for `uniqid()`?

erik-konrad replied:
Thats a suggestion from phpstorm for uniqid().

Commit:
[FIX] dynamic timestamp for starttime and endtime in queries

sebastian-meyer commented on Classes/Plugin/OaiPmh.php:
```diff
@@ -449,35 +456,43 @@ protected function resume()
+            'FROM `tx_dlf_documents` ' .
+            'INNER JOIN `tx_dlf_relations` ON `tx_dlf_relations`.`uid_local` = `tx_dlf_documents`.`uid` ' .
+            'INNER JOIN `tx_dlf_collections` ON `tx_dlf_collections`.`uid` = `tx_dlf_relations`.`uid_foreign` ' .
+            'WHERE (((`tx_dlf_documents`.`deleted` = 0) AND (`tx_dlf_collections`.`deleted` = 0)) AND ((`tx_dlf_documents`.`hidden` = 0) AND (`tx_dlf_collections`.`hidden` = 0)) AND (`tx_dlf_documents`.`starttime` <= ' . $GLOBALS['EXEC_TIME'] . ') AND ((`tx_dlf_documents`.`endtime` = 0) OR (`tx_dlf_documents`.`endtime` > ' . $GLOBALS['EXEC_TIME'] . '))) ' .
```
This should use `Helper::whereExpression()` to handle all the `enableFields` correctly. Attention: `whereExpression()` returns a clause starting with `AND`, so it must not be the first where clause.

sebastian-meyer commented on Classes/Plugin/OaiPmh.php:
```diff
@@ -449,35 +456,43 @@ protected function resume()
+            'AND tx_dlf_collections.pid = ? ' .
+            'AND tx_dlf_relations.ident="docs_colls"' .
+            $where .
+            Helper::whereClause('tx_dlf_collections', [$this->piVars['identifier']]);
```
The second argument of `Helper::whereClause()` has to be a boolean (or can be omitted in this case). It defines if hidden records should be shown.

Also, in order to use the QueryBuilder, using `Helper::whereExpression()` is prefered over using `Helper::whereClause()` (which uses the old DB API functions).

albig replied:
I disagree on using `Helper::whereClause()` at all. It adds similar restriction constraints as QueryBuilder did already. This results is an unneeded complex SQL-query IMHO.

Example of a non working (other problem in line 795) Query with and without  `Helper::whereClause()`:

### with  `Helper::whereClause()`
<pre>
SELECT `tx_dlf_collections`.`oai_name` AS `oai_name`, `tx_dlf_collections`.`label` AS `label` 
FROM `tx_dlf_collections` 
WHERE (`tx_dlf_collections`.`sys_language_uid` IN (-1, 0)) 
AND (`tx_dlf_collections`.`pid` = 176) 
AND (`tx_dlf_collections`.`oai_name` <> ) 
AND (`tx_dlf_collections`.`fe_cruser_id` = 0) 
AND (
  (`tx_dlf_collections`.`deleted` = 0) 
  AND (`tx_dlf_collections`.`hidden` = 0) 
  AND (
    ( 
      (`tx_dlf_collections`.`fe_group` = '') 
      OR (`tx_dlf_collections`.`fe_group` IS NULL) 
      OR (`tx_dlf_collections`.`fe_group` = '0') 
      OR (FIND_IN_SET('0', `tx_dlf_collections`.`fe_group`))
      OR (FIND_IN_SET('-1', `tx_dlf_collections`.`fe_group`)) 
     )
   )
)
AND (
  (`tx_dlf_collections`.`deleted` = 0) 
  AND (`tx_dlf_collections`.`hidden` = 0)
 )
ORDER BY `tx_dlf_collections`.`oai_name` ASC
</pre>

### without  `Helper::whereClause()`
<pre>
SELECT `tx_dlf_collections`.`oai_name` AS `oai_name`, `tx_dlf_collections`.`label` AS `label` 
FROM `tx_dlf_collections` 
WHERE (`tx_dlf_collections`.`sys_language_uid` IN (-1, 0)) 
AND (`tx_dlf_collections`.`pid` = 176) 
AND (`tx_dlf_collections`.`oai_name` <> ) 
AND (`tx_dlf_collections`.`fe_cruser_id` = 0) 
AND (
  (`tx_dlf_collections`.`deleted` = 0) 
  AND (`tx_dlf_collections`.`hidden` = 0)
) 
ORDER BY `tx_dlf_collections`.`oai_name` ASC
</pre>

sebastian-meyer commented on Classes/Plugin/OaiPmh.php:
```diff
@@ -931,24 +946,32 @@ protected function fetchDocumentUIDs()
+            'FROM `tx_dlf_documents` ' .
+            'INNER JOIN `tx_dlf_relations` ON `tx_dlf_relations`.`uid_local` = `tx_dlf_documents`.`uid` ' .
+            'INNER JOIN `tx_dlf_collections` ON `tx_dlf_collections`.`uid` = `tx_dlf_relations`.`uid_foreign` ' .
+            'WHERE (((`tx_dlf_documents`.`deleted` = 0) AND (`tx_dlf_collections`.`deleted` = 0)) AND ((`tx_dlf_documents`.`hidden` = 0) AND (`tx_dlf_collections`.`hidden` = 0)) AND (`tx_dlf_documents`.`starttime` <= ' . $GLOBALS['EXEC_TIME'] . ') AND ((`tx_dlf_documents`.`endtime` = 0) OR (`tx_dlf_documents`.`endtime` > ' . $GLOBALS['EXEC_TIME'] . '))) ' .
```
See comment above:
This should use `Helper::whereExpression()` to handle all the `enableFields` correctly. Attention: `whereExpression()` returns a clause starting with `AND`, so it must not be the first where clause.

sebastian-meyer commented on Classes/Plugin/OaiPmh.php:
```diff
@@ -931,24 +946,32 @@ protected function fetchDocumentUIDs()
+            'AND tx_dlf_documents.pid = ? ' .
+            'AND tx_dlf_collections.pid = ? ' .
+            'AND tx_dlf_relations.ident="docs_colls" ' .
+            Helper::whereClause('tx_dlf_collections') .
```
See comment above:
In order to use the QueryBuilder, using Helper::whereExpression() is prefered over using Helper::whereClause() (which uses the old DB API functions).

albig commented on Classes/Plugin/OaiPmh.php:
```diff
@@ -931,24 +946,32 @@ protected function fetchDocumentUIDs()
+
         $output = $this->oai->createElementNS('http://www.openarchives.org/OAI/2.0/', $verb);
-        while ($resArray = $GLOBALS['TYPO3_DB']->sql_fetch_assoc($documents)) {
+        while ($resArray = $documents->fetch()) {
```
Actually the $resArray() is an array with only `NULL` values. So the query above must be wrong on calling 
 
[verb=ListRecords&metadataPrefix=mets](url)

<pre>
array(41 items)
   uid => NULL
   pid => NULL
   tstamp => NULL
   crdate => NULL
   cruser_id => NULL
   deleted => NULL
   hidden => NULL
   starttime => NULL
   endtime => NULL
   fe_group => NULL
   prod_id => NULL
   location => NULL
   record_id => NULL
   opac_id => NULL
   union_id => NULL
   urn => NULL
   purl => NULL
   title => NULL
   title_sorting => NULL
   author => NULL
   year => NULL
   place => NULL
   thumbnail => NULL
   metadata => NULL
   metadata_sorting => NULL
   structure => NULL
   partof => NULL
   volume => NULL
   volume_sorting => NULL
   license => NULL
   terms => NULL
   restrictions => NULL
   out_of_print => NULL
   rights_info => NULL
   collections => NULL
   mets_label => NULL
   mets_orderlabel => NULL
   owner => NULL
   solrcore => NULL
   status => NULL
   document_format => NULL
</pre>

The same with

[?verb=ListIdentifiers&metadataPrefix=oai_dc](url)
[?verb=ListIdentifiers&metadataPrefix=mets](url)
[?verb=ListIdentifiers&metadataPrefix=epicur](url)
[?verb=ListRecords&metadataPrefix=oai_dc](url)
[?verb=ListRecords&metadataPrefix=epicur](url)

albig replied:
This is partly fixed. The array contains values. But the query is still somehow wrong. There is always only one document found.


**Example**

`$documentsToProcess` containts the UIDs 1 to 10.

```
SELECT `tx_dlf_documents`.*, GROUP_CONCAT(DISTINCT `tx_dlf_collections`.`oai_name` ORDER BY `tx_dlf_collections`.`oai_name` SEPARATOR " ") AS `collections`              FROM `tx_dlf_documents`              INNER JOIN `tx_dlf_relations` ON `tx_dlf_relations`.`uid_local` = `tx_dlf_documents`.`uid`              INNER JOIN `tx_dlf_collections` ON `tx_dlf_collections`.`uid` = `tx_dlf_relations`.`uid_foreign`              WHERE `tx_dlf_documents`.`uid` IN ( 1,2,3,4,5,6,7,8,9,10 ) AND `tx_dlf_relations`.`ident`="docs_colls";
```
**Result (shorted)**

```
*************************** 1. row ***************************
             uid: 1
             pid: 12
          tstamp: 1569320630
          crdate: 1564134023
       cruser_id: 2
         deleted: 0
          hidden: 0
       starttime: 0
         endtime: 0
        fe_group: 
         prod_id: 336246
        location: https://example.com/00052637/00052637_mets.xml
       record_id: 00052637
         opac_id: 
        union_id: 
           title: Title of Document
          author: 
            year: 
           place: 
       structure: 33
          partof: 0
          volume: 
  volume_sorting: 
     collections: 1
           owner: 1
        solrcore: 1
          status: 0
 document_format: METS
         license: 
           terms: 
    restrictions: 
    out_of_print: 
     rights_info: 
      mets_label: 
 mets_orderlabel: 

1 row in set (0.001 sec)

```

albig reviewed (changes requested):
I abort the review and first try to make the current OAI-plugin working with TYPO 8.7 LTS (which should do already).

Commit:
[FIX] Query refactoring. removes static expressions and whereClause calls and sets whereExpression instead.

sebastian-meyer commented on Classes/Plugin/OaiPmh.php:
```diff
@@ -449,35 +456,42 @@ protected function resume()
+
+        $connection = GeneralUtility::makeInstance(ConnectionPool::class)->getConnectionForTable('tx_dlf_documents');
+
+        $sql = 'SELECT `tx_dlf_documents`.*, GROUP_CONCAT(DISTINCT `tx_dlf_collections`.oai_name ORDER BY tx_dlf_collections.oai_name SEPARATOR " ") AS `collections` ' .
```
Please consider using table and field enclosing consistently. Currently some table names are enclosed in backticks and some are not. Generally we should enclose all table and field names in backticks to be safe.

erik-konrad replied:
done

sebastian-meyer commented on Classes/Plugin/OaiPmh.php:
```diff
@@ -449,35 +456,42 @@ protected function resume()
+            'AND tx_dlf_collections.pid = ? ' .
+            'AND tx_dlf_relations.ident="docs_colls" ' .
+            $where .
+            'AND ' . Helper::whereExpression('tx_dlf_collections', [$this->piVars['identifier']]);
```
The second parameter of `Helper::whereExpression()` has to be of type `bool` and should be `false` or omitted (since this is the default value) in this case.

erik-konrad replied:
done

sebastian-meyer commented on Classes/Plugin/OaiPmh.php:
```diff
@@ -931,24 +945,31 @@ protected function fetchDocumentUIDs()
+
+        $connection = GeneralUtility::makeInstance(ConnectionPool::class)->getConnectionForTable('tx_dlf_documents');
+
+        $sql = 'SELECT `tx_dlf_documents`.*, GROUP_CONCAT(DISTINCT `tx_dlf_collections`.oai_name ORDER BY tx_dlf_collections.oai_name SEPARATOR " ") AS `collections` ' .
```
See above: Every table and field name should be enclosed in backticks.

erik-konrad replied:
done too

sebastian-meyer committed:
Remove newline for better code styling

sebastian-meyer committed:
Merge branch 'master' into master

Commit:
[FIX] code review fixes

sebastian-meyer reviewed (dismissed):
Thank you very much! I approve merging these changes, now.

What about the change request from @albig in https://github.com/kitodo/kitodo-presentation/pull/503#discussion_r410163677?

sebastian-meyer committed:
Merge branch 'master' into master

sebastian-meyer committed:
Merge branch 'master' into master

sebastian-meyer commented:
@albig You have requested changes in https://github.com/kitodo/kitodo-presentation/pull/503#discussion_r410163677. Is this still valid?

sebastian-meyer committed:
Merge branch 'master' into master

albig commented:
I haven't looked at it anymore. Is it rebased already to the current master?

sebastian-meyer committed:
Merge branch 'master' into master

sebastian-meyer commented:
Yes, this branch is up-to-date.

Commit:
Fix doctrine prepare statement.

    The current solution only works for the first uid in the list of
    $documentsToProcess. This is due to a wrong usage of the doctrine
    prepare statement.

    Additionally the GROUP BY is required in the SQL query. Otherwise only
    one result is get.

Commit:
bring back whereExpression() in generateOutputForDocumentList

erik-konrad committed:
Merge pull request #1 from albig/erik-konrad-master

Fix doctrine prepare statement.

sebastian-meyer commented on Classes/Plugin/OaiPmh.php:
```diff
@@ -958,14 +960,25 @@ protected function generateOutputForDocumentList(DocumentList $documentListSet)
-        $documents->bindValue(2, (int)$this->conf['pages']);
-        $documents->bindValue(3, (int)$this->conf['pages']);
-        $documents->bindValue(4, (int)$this->conf['limit']);
         $documents->execute();
```
```suggestion
```
`$connection->executeQuery()` already returns a database result object, so we don't need another `execute()` here.

Commit:
Update Classes/Plugin/OaiPmh.php

You're right. This is not necessary anymore.

Co-authored-by: Sebastian Meyer <sebastian.meyer@slub-dresden.de>

albig commented on Classes/Plugin/OaiPmh.php:
```diff
@@ -449,35 +456,42 @@ protected function resume()
+            $where .
+            'AND ' . Helper::whereExpression('tx_dlf_collections');
+
+        $statement = $connection->prepare($sql);
```
Could you rewrite the `prepare()` to `executeQuery()` to use the same code pattern as in `generateOutputForDocumentList()`?

This code is working as there is no `IN()`-operator in use. But nevertheless it will be confusing on next code rewriting.

erik-konrad replied:
done

albig reviewed (changes requested):
That's the last change request. After this we can merge this PR.

sebastian-meyer committed:
Fix Codacy Issue

sebastian-meyer committed:
Merge branch 'master' into master

Commit:
refactores $connection-prepare to $connection->executeQuery

Commit:
Merge branch 'master' of https://github.com/erik-konrad/kitodo-presentation

sebastian-meyer committed:
Use statements in alphabetical order

sebastian-meyer committed:
Merge branch 'master' into master

albig reviewed (approved):
Thank you. Seems to work now with TYPO3 9.5 LTS.

sebastian-meyer commented:
@erik-konrad Thank you very much for your contribution!
